File: skills/actor-pattern/zig-compendium/02_request_response.zig
// Example 02: Request-Response Actor — Key-Value Store
//
// A more realistic actor: an in-memory key-value store with set, get, delete,
// and keys operations. Demonstrates:
//   - Multiple payload fields on the command union
//   - Returning copies of data (not references to actor-owned state)
//   - Error handling within the actor (key not found)
//   - Tagged union variants with different reply channel types
//
// Build & run:
//   cd zig-compendium && zig build-exe 02_request_response.zig && ./02_request_response

const std = @import("std");
const Channel = @import("channel.zig").Channel;
const print = std.debug.print;

const allocator = std.heap.page_allocator;

// --- Reply types ---

const GetReply = struct {
    value: ?[]const u8, // null means not found
};

const KeysReply = struct {
    keys: [][]const u8, // caller must free
};

// --- Command definition ---

const KVCmd = union(enum) {
    set: struct {
        key: []const u8,
        value: []const u8,
        reply: *Channel(void),
    },
    get: struct {
        key: []const u8,
        reply: *Channel(GetReply),
    },
    delete: struct {
        key: []const u8,
        reply: *Channel(void),
    },
    keys: struct {
        reply: *Channel(KeysReply),
    },
};

// --- Actor: dispatch thread ---

fn kvActor(cmds: *Channel(KVCmd)) void {
    var data = std.StringHashMap([]const u8).init(allocator);
    defer data.deinit();

    while (cmds.recv()) |cmd| {
        switch (cmd) {
            .set => |c| {
                data.put(c.key, c.value) catch {};
                c.reply.send({});
            },
            .get => |c| {
                const val = data.get(c.key);
                c.reply.send(.{ .value = val });
            },
            .delete => |c| {
                _ = data.remove(c.key);
                c.reply.send({});
            },
            .keys => |c| {
                var key_list = std.ArrayList([]const u8).initCapacity(allocator, data.count()) catch unreachable;
                var it = data.keyIterator();
                while (it.next()) |k| {
                    key_list.append(allocator, k.*) catch {};
                }
                c.reply.send(.{ .keys = key_list.toOwnedSlice(allocator) catch &.{} });
            },
        }
    }
}

// --- Convenience functions ---

fn kvSet(cmds: *Channel(KVCmd), key: []const u8, value: []const u8) void {
    var reply = Channel(void).init(allocator, 1);
    defer reply.deinit();
    cmds.send(.{ .set = .{ .key = key, .value = value, .reply = &reply } });
    _ = reply.recv();
}

fn kvGet(cmds: *Channel(KVCmd), key: []const u8) ?[]const u8 {
    var reply = Channel(GetReply).init(allocator, 1);
    defer reply.deinit();
    cmds.send(.{ .get = .{ .key = key, .reply = &reply } });
    const r = reply.recv() orelse return null;
    return r.value;
}

fn kvDelete(cmds: *Channel(KVCmd), key: []const u8) void {
    var reply = Channel(void).init(allocator, 1);
    defer reply.deinit();
    cmds.send(.{ .delete = .{ .key = key, .reply = &reply } });
    _ = reply.recv();
}

fn kvKeys(cmds: *Channel(KVCmd)) [][]const u8 {
    var reply = Channel(KeysReply).init(allocator, 1);
    defer reply.deinit();
    cmds.send(.{ .keys = .{ .reply = &reply } });
    const r = reply.recv() orelse return &.{};
    return r.keys;
}

// --- Demo ---

pub fn main() !void {
    var cmds = Channel(KVCmd).init(allocator, 64);
    defer cmds.deinit();

    const actor = try std.Thread.spawn(.{}, kvActor, .{&cmds});

    // Set values from multiple threads.
    const pairs = [_]struct { k: []const u8, v: []const u8 }{
        .{ .k = "name", .v = "Alice" },
        .{ .k = "city", .v = "Portland" },
        .{ .k = "lang", .v = "Zig" },
        .{ .k = "editor", .v = "Emacs" },
    };

    var workers: [pairs.len]std.Thread = undefined;
    for (&pairs, 0..) |p, i| {
        workers[i] = try std.Thread.spawn(.{}, struct {
            fn run(c: *Channel(KVCmd), key: []const u8, val: []const u8) void {
                kvSet(c, key, val);
            }
        }.run, .{ &cmds, p.k, p.v });
    }
    for (&workers) |*w| w.join();

    // Read them back.
    for (&pairs) |p| {
        if (kvGet(&cmds, p.k)) |v| {
            print("{s} = {s}\n", .{ p.k, v });
        } else {
            print("{s} = (not found)\n", .{p.k});
        }
    }

    // Try a missing key.
    if (kvGet(&cmds, "nonexistent")) |_| {
        print("nonexistent = (unexpected)\n", .{});
    } else {
        print("missing key: \"nonexistent\" not found\n", .{});
    }

    // Delete and list keys.
    kvDelete(&cmds, "city");
    const keys = kvKeys(&cmds);
    defer allocator.free(keys);

    print("keys after delete:", .{});
    for (keys) |k| {
        print(" {s}", .{k});
    }
    print("\n", .{});

    // Shut down.
    cmds.close();
    actor.join();
}

File: skills/actor-pattern/rust-compendium/src/bin/02_request_response.rs
//! Example 02: Request-Response Actor -- Key-Value Store
//!
//! A more realistic actor: an in-memory key-value store with Set, Get, Delete,
//! and Keys operations. Demonstrates:
//!   - Multiple payload fields embedded in command variants
//!   - Returning clones of data (not references to actor-owned state)
//!   - Error handling within the actor (key not found)
//!   - The handle struct is Clone-able for sharing across threads
//!
//! Run: cargo run --bin 02_request_response

use std::collections::HashMap;
use std::sync::mpsc::{self, Sender};
use std::thread;

// --- Command enum ---

enum Command {
    Set {
        key: String,
        value: String,
        reply: Sender<()>,
    },
    Get {
        key: String,
        reply: Sender<Result<String, String>>,
    },
    Delete {
        key: String,
        reply: Sender<()>,
    },
    Keys {
        reply: Sender<Vec<String>>,
    },
}

// --- Actor handle ---

#[derive(Clone)]
struct KVStore {
    tx: Sender<Command>,
}

impl KVStore {
    fn new() -> Self {
        let (tx, rx) = mpsc::channel::<Command>();

        thread::spawn(move || {
            // All mutable state lives here.
            let mut data: HashMap<String, String> = HashMap::new();

            while let Ok(cmd) = rx.recv() {
                match cmd {
                    Command::Set { key, value, reply } => {
                        data.insert(key, value);
                        let _ = reply.send(());
                    }
                    Command::Get { key, reply } => {
                        let result = match data.get(&key) {
                            Some(v) => Ok(v.clone()),
                            None => Err(format!("key not found: {:?}", key)),
                        };
                        let _ = reply.send(result);
                    }
                    Command::Delete { key, reply } => {
                        data.remove(&key);
                        let _ = reply.send(());
                    }
                    Command::Keys { reply } => {
                        let mut keys: Vec<String> = data.keys().cloned().collect();
                        keys.sort();
                        let _ = reply.send(keys);
                    }
                }
            }
        });

        KVStore { tx }
    }

    fn set(&self, key: &str, value: &str) {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx
            .send(Command::Set {
                key: key.to_string(),
                value: value.to_string(),
                reply: reply_tx,
            })
            .unwrap();
        reply_rx.recv().unwrap();
    }

    fn get(&self, key: &str) -> Result<String, String> {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx
            .send(Command::Get {
                key: key.to_string(),
                reply: reply_tx,
            })
            .unwrap();
        reply_rx.recv().unwrap()
    }

    fn delete(&self, key: &str) {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx
            .send(Command::Delete {
                key: key.to_string(),
                reply: reply_tx,
            })
            .unwrap();
        reply_rx.recv().unwrap();
    }

    fn keys(&self) -> Vec<String> {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx.send(Command::Keys { reply: reply_tx }).unwrap();
        reply_rx.recv().unwrap()
    }
}

// --- Demo ---

fn main() {
    let store = KVStore::new();

    // Set some values from multiple threads.
    let pairs = vec![
        ("name", "Alice"),
        ("city", "Portland"),
        ("lang", "Rust"),
        ("editor", "Emacs"),
    ];

    let mut handles = Vec::new();
    for (k, v) in &pairs {
        let s = store.clone();
        let key = k.to_string();
        let val = v.to_string();
        handles.push(thread::spawn(move || {
            s.set(&key, &val);
        }));
    }
    for h in handles {
        h.join().unwrap();
    }

    // Read them back.
    for (k, _) in &pairs {
        match store.get(k) {
            Ok(v) => println!("{} = {}", k, v),
            Err(e) => println!("get {:?}: error: {}", k, e),
        }
    }

    // Try a missing key.
    match store.get("nonexistent") {
        Ok(v) => println!("nonexistent = {}", v),
        Err(e) => println!("missing key error: {}", e),
    }

    // Delete and verify.
    store.delete("city");
    let keys = store.keys();
    println!("keys after delete: {:?}", keys);
}
